File: 2025/aoc/day07/puzzle.py
from functools import cache

import aoc.lib.constants as const
from aoc.lib.location import Location
from aoc.lib.puzzle import Puzzle


class Node:
    TYPE_START = "S"
    TYPE_SPLITTER = "^"
    TYPE_EMPTY = "."
    TYPE_BEAM = "|"

    def __init__(self, ntype, location):
        self.type = ntype
        self.loc = location

        self.paths = []

    def __str__(self):
        return f"Node({self.type},{self.loc})"

    def __eq__(self, other):
        return self.loc == other.loc

    def __hash__(self):
        return hash(self.loc)

    def add_path(self, node):
        self.paths.append(node)


class Laboratories(Puzzle):
    """AOC-2025 // Day07 -- Laboratories"""

    def __init__(self, input_file, **kwargs):
        super().__init__(input_file, **kwargs)

        self.__tachyon_grid = []
        self._read_input(self.__parse_input)

        start_r = 0
        start_c = self.__tachyon_grid[0].index(Node.TYPE_START)
        self.__start_loc = Location(start_r, start_c)
        self.__exit_row = len(self.__tachyon_grid) - 1

    def __parse_input(self, line):
        self.__tachyon_grid.append(list(line))

    def _part1(self):
        self._debug(f"Start Loc: {self.__start_loc}")

        split_count = 0
        beams = set()
        beams.add(self.__start_loc.nearby(const.S))
        while beams:
            self._debug(f"  -> Beam Count: [{len(beams)}]")
            beams_to_add = set()
            beams_to_remove = set()
            for beam in beams:
                if beam.row != self.__exit_row:
                    # look south of beam to see what's there
                    next_loc = beam.nearby(const.S)
                    next_space = self.__tachyon_grid[next_loc.row][next_loc.col]

                    if next_space == Node.TYPE_SPLITTER:
                        split_count += 1
                        self._debug(f"  -> Split Count: [{split_count}]")

                        # Split the beam into two new beams
                        new_beam1 = beam.nearby(const.SW)
                        beams_to_add.add(new_beam1)

                        new_beam2 = beam.nearby(const.SE)
                        beams_to_add.add(new_beam2)
                    elif next_space == Node.TYPE_EMPTY:
                        # Move down/S
                        new_loc = beam.nearby(const.S)
                        # Add new beam pos
                        beams_to_add.add(new_loc)

                # Remove the old beam
                beams_to_remove.add(beam)

            beams.difference_update(beams_to_remove)
            beams.update(beams_to_add)

        return split_count

    @cache  # noqa: B019
    def __traverse_manifold(self, curr_node):
        timelines = 0
        new_nodes = []

        match curr_node.type:
            case Node.TYPE_START | Node.TYPE_EMPTY:
                loc = curr_node.loc.nearby(const.S)
                ntype = self.__tachyon_grid[loc.row][loc.col]
                node = Node(ntype, loc)
                new_nodes.append(node)
            case Node.TYPE_SPLITTER:
                timelines += 1
                self._debug(f"-> Timelines: {timelines}")
                for direction in (const.SE, const.SW):
                    loc = curr_node.loc.nearby(direction)
                    ntype = self.__tachyon_grid[loc.row][loc.col]
                    node = Node(ntype, loc)
                    new_nodes.append(node)

        for node in new_nodes:
            if node.loc.row < self.__exit_row:
                curr_node.add_path(node)
                timelines += self.__traverse_manifold(node)
            else:
                self._debug(f"-> Exit at {node}")

        return timelines

    def _part2(self):
        start_node = Node(Node.TYPE_START, self.__start_loc)
        self._debug(start_node)

        # Start in the current timeline
        timelines = 1

        # Also builds a tree of nodes, but it's not used
        timelines += self.__traverse_manifold(start_node)

        return timelines

File: 2025/agoc/src/caroon.org/aoc25/day07/puzzle.go
package day07

import (
	"fmt"
	"slices"
	"strings"

	"caroon.org/aoc25/shared/grid"
)

const DayName string = "Day07"
const Title string = "Laboratories"

const startMarker string = "S"
const splitterMarker string = "^"
const emptyMarker string = "."

type Node struct {
	marker string
	loc    grid.Location
}

func solvePart1(tachyonManifold [][]string) int {
	var splitCount int = 0
	var exitRow int = len(tachyonManifold) - 1
	// Used to keep track of the active Beam locations
	// Don't really care about the value, just need the
	// uniqueness of the keys
	// Used in lieu of a mathematical set
	var beams map[grid.Location]bool = map[grid.Location]bool{}

	startCol := slices.Index(tachyonManifold[0], startMarker)
	startLoc := grid.Location{Row: 0, Col: startCol}
	// Add the first beam to get started
	beams[startLoc] = true

	for len(beams) > 0 {
		beamsAdded := make(map[grid.Location]bool)
		beamsRemoved := make(map[grid.Location]bool)

		for beam, _ := range beams {
			if beam.Row != exitRow {
				// look south of beam to see what's there
				nextLoc := beam.Nearby(grid.S)
				nextSpace := tachyonManifold[nextLoc.Row][nextLoc.Col]

				if nextSpace == splitterMarker {
					splitCount += 1
					// Split the beam into two new beams
					newBeam1 := beam.Nearby(grid.SW)
					beamsAdded[newBeam1] = true

					newBeam2 := beam.Nearby(grid.SE)
					beamsAdded[newBeam2] = true

				} else if nextSpace == emptyMarker {
					// Move down/S
					newLoc := beam.Nearby(grid.S)
					// Add new beam pos
					beamsAdded[newLoc] = true
				}
			}
			// Remove the old beam
			beamsRemoved[beam] = true
		}

		// removed dead beams
		for loc, _ := range beamsRemoved {
			delete(beams, loc)
		}

		// add new beams
		for loc, _ := range beamsAdded {
			beams[loc] = true
		}

	}

	return splitCount
}

func traverseManifold(tachyonManifold [][]string, currNode Node, cache map[Node]int) int {
	var timelines int = 0
	newNodes := []Node{}

	if cache[currNode] > 0 {
		timelines = cache[currNode]
	} else {
		switch currNode.marker {
		case startMarker:
			fallthrough
		case emptyMarker:
			loc := currNode.loc.Nearby(grid.S)
			marker := tachyonManifold[loc.Row][loc.Col]
			node := Node{marker: marker, loc: loc}
			newNodes = append(newNodes, node)
		case splitterMarker:
			timelines += 1
			for _, direction := range []grid.Direction{grid.SE, grid.SW} {
				loc := currNode.loc.Nearby(direction)
				marker := tachyonManifold[loc.Row][loc.Col]
				node := Node{marker: marker, loc: loc}
				newNodes = append(newNodes, node)
			}
		}

		for _, node := range newNodes {
			if node.loc.Row < len(tachyonManifold)-1 {
				timelines += traverseManifold(tachyonManifold, node, cache)
			}
		}
		cache[currNode] = timelines
	}

	return timelines
}

func solvePart2(tachyonManifold [][]string) int {
	// Start in current timeline ... of which there is 1
	var timelines int = 1
	// For memoization
	var cache map[Node]int = make(map[Node]int)

	startCol := slices.Index(tachyonManifold[0], startMarker)
	startLoc := grid.Location{Row: 0, Col: startCol}
	startNode := Node{marker: startMarker, loc: startLoc}

	// cache -> to Memoize the traverseManfold function
	timelines += traverseManifold(tachyonManifold, startNode, cache)

	return timelines
}

func processInput(data []string) [][]string {
	var tachyonManifold [][]string

	for _, line := range data {
		markers := strings.Split(line, "")
		tachyonManifold = append(tachyonManifold, markers)
	}

	return tachyonManifold
}

func Exec(part string, data []string) (int, error) {
	var result int = 0
	var err error = nil

	tachyonManifold := processInput(data)

	if part == "PART1" {
		result = solvePart1(tachyonManifold)
	} else if part == "PART2" {
		result = solvePart2(tachyonManifold)
	} else {
		err = fmt.Errorf("%s - Unknown Part: [%s]\n", DayName, part)
	}

	if result == -42 {
		err = fmt.Errorf("%s - %s Not Implemented\n", DayName, part)
	}

	return result, err
}
